docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,89 @@
+11. Enable producing to event bus via settings
+##############################################
+
+Status
+******
+
+**Accepted** 2023-06-05
+
+Context
+*******
+
+The initial implementation of the event bus allowed only a single event type to be published to a topic, with details like topic name, consumer group name, and consumer name configured via code. We weren't sure what the api would finally look like, and whether the event bus config would live as part of the definition of the event, so we just went with explicit code that we could iterate upon.
+
+The current implementation of openedx-events does not actually push any events to the underlying implementations like `edx-event-bus-kafka`_ and `edx-event-bus-redis`_. The event-producing application is expected to create a signal handler (since openedx-events subclasses Django signals) to catch the event and push it into the event bus. Some examples of the handlers: `handlers example`_.
+
+.. _handlers example: https://github.com/openedx/edx-platform/blob/27b8d2f68d5dfaf84755e7d7f8dccc97ce3be509/cms/djangoapps/contentstore/signals/handlers.py#L162-L210
+.. _edx-event-bus-kafka: https://github.com/openedx/event-bus-kafka
+.. _edx-event-bus-redis: https://github.com/openedx/event-bus-redis
+
+This ADR aims to propose a solution for configuring the details, like the topic name, consumer group name, etc. via Django settings as well as pushing events to the event bus without requiring the producing application to write additional handlers.
+
+
+Decision
+********
+
+Create a generic signal handler to push events to the event bus. This handler should be attached to or connected to the signals that are enabled in Django settings. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   # .. setting_name: EVENT_BUS_PRODUCER_CONFIG
+   # .. setting_default: {}
+   # .. setting_description: Dictionary specifying event_types (keys) and its configuration dictionary as values.
+   #   Each configuration dictionary contains a flag denoting whether the event will be enabled or disabled,
+   #   topics/stream names where the event will be pushed to, event_key_field which is
+   #   and period delimited string path to event data field to use as event key.
+   #   Note: The topic names should not include environment prefix as it will be dynamically added based on
+   #   EVENT_BUS_TOPIC_PREFIX setting.
```

Comment:
can we add a code inline annotation that specifies each field for the configurations? It looks a bit cluttered if it's part of the description.

Reply:
There, we could specify the defaults if any

Reply:
@mariajgrimaldi The configuration part has changed a little. I am not sure on how to go about adding annotation inline inside the dictionary. Currently, it is like this:

```python
# .. setting_name: EVENT_BUS_PRODUCER_CONFIG
# .. setting_default: {}
# .. setting_description: Dictionary with event_types as keys and list of dictonaries for topic related configuration.
#    Each topic configuration dictionary contains a flag denoting whether the event will be published to the topic,
#    topic/stream name where the event will be pushed to, event_key_field which is and period delimited string path
#    to event data field to use as event key.
#    Note: The topic names should not include environment prefix as it will be dynamically added based on
#    EVENT_BUS_TOPIC_PREFIX setting.
EVENT_BUS_PRODUCER_CONFIG = {
    'org.openedx.content_authoring.xblock.published.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
        {'topic': 'content-authoring-xblock-published', 'event_key_field': 'xblock_info.usage_key', 'enabled': False},
    ],
    'org.openedx.content_authoring.xblock.deleted.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
    ],
}
```

By adding code inline annotation, do you mean some thing like:

```python
   # .. setting_name: EVENT_BUS_PRODUCER_CONFIG
   # .. setting_default: {}
   # .. setting_description: Dictionary with event_types as keys and list of dictonaries for topic related configuration.
   EVENT_BUS_PRODUCER_CONFIG = {
       'org.openedx.content_authoring.xblock.published.v1': [
        # .. setting_default: {}
        #    Each topic configuration dictionary contains a flag denoting whether the event will be published to the topic,
        #    topic/stream name where the event will be pushed to, event_key_field which is and period delimited string path
        #    to event data field to use as event key.
        #    Note: The topic names should not include environment prefix as it will be dynamically added based on
        #    EVENT_BUS_TOPIC_PREFIX setting.
           {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
           {'topic': 'xblock-published', 'event_key_field': 'xblock_info.usage_key', 'enabled': False},
       ],
       'org.openedx.content_authoring.xblock.deleted.v1': [
           {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
       ],
   }

```

Reply:
When the annotations are extracted to an RST doc:
1. All the annotations must remain together to be parsed correctly.
2. The code will not be available in the doc, just the annotation text.
3. We can't invent new annotations without annotation definition changes that make them recognizable.

We can do something like the following though:
```
# .. setting_description: Dictionary with event_types as keys and a list of dictionaries for topic related configuration. 
#    Each dictionary in the list contains:
#      topic: <explain this>.
#      event_key_field: <explain this>.
#      ...
```
Note that the final RST has no formatting, so these indents would be lost if this were ever turned into a doc.

Reply:
I'm pretty happy with what we have now -- I think it just needs explicit naming of the fields (`enabled`, `topic`) and it's ready to go, as far as I'm concerned.

Reply:
Updated with explicit names in description.